---
title: "Axial Decontam 2023"
author: "Alexis Adams"
date: "2025-08-24"
output: html_document
---

#Visualize Data

Check how many sequences and ASVs are in each sample. 

Original ASV table has ASV featureID as rows and samples are columns. To make the figure, make samples = rows and FeatureID = column. Mutate new column for sum of sequences. To do this, take the column "X.OTU.ID" and make it the rownames. t() transposes (flips) rows the columns. After transposing, the object becomes a matrix, so convert back to DF. Rename X.OTU.ID to sample. 
```{r}
ASV <- read.delim("C:/Users/alexis98adams/Documents/R/Axial2023/datainput/deepsea-18s-merged-asv-table-05022025.tsv", header = TRUE, sep = "\t", skip = 1)

Rearranged_asv_axial <- ASV %>%
  column_to_rownames("X.OTU.ID") %>%  
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  mutate(Seq_sum = rowSums(across(-sample))) %>% 
  filter(grepl("Axial", sample, ignore.case = TRUE))

```

#Number of sequences in each sample. Use a bar plot to see exact counts for each sample (histogram would just show distribution)
```{r}
ggplot(Rearranged_asv_axial, aes(x= sample, y = Seq_sum)) +
  geom_col(fill = "steelblue") +
  labs(x = "Sample", y = "Number of sequences") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

```
#Number of ASVs in each sample
```{r}
Rearranged_asv_axial <- Rearranged_asv_axial %>% rowwise() %>% 
  mutate(ASV_sum = sum(c_across(-sample) > 0, na.rm = TRUE)) %>%
  ungroup()

ggplot(Rearranged_asv_axial, aes(x= sample, y = ASV_sum)) +
  geom_col(fill = "coral") +
  labs(x = "Sample", y = "Number of ASVs in each sample") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

```
#Run decontam. Create new data table and asv table that only contains the sequences that passed decontam. 
```{r}
library(tidyverse)
install.packages("phyloseq")
library(phyloseq)
library(dplyr)

if(!requireNamespace("BiocManager")){
  install.packages("BiocManager")
}
BiocManager::install("phyloseq")

if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("decontam")

library(decontam)
library(phyloseq)

#Read in data

ASV <- read.delim("C:/Users/alexis98adams/Documents/R/Axial2023/datainput/deepsea-18s-merged-asv-table-05022025.tsv", header = TRUE, sep = "\t", skip = 1)
tax <- read.delim("C:/Users/alexis98adams/Documents/R/Axial2023/datainput/taxonomy.tsv", header = TRUE, sep = "\t")
metadata <- read.csv("C:/Users/alexis98adams/Documents/R/Axial2023/datainput/axial_compiledmetadata.csv", header = TRUE)

#Filter ASV table for just Axial Samples

axial_asv <- ASV %>% select(Feature.ID = 'X.OTU.ID', everything()) %>%
  pivot_longer(cols = !Feature.ID,
               names_to = "sample", values_to = "value") %>%
  filter(grepl("Axial", sample, ignore.case = TRUE))

#Remove 16S samples from metadata
metadata <- metadata %>% filter(PRIMER != "16S") %>% rename(sample = NEW_FASTQ_NAME)

axial_asv_tax_metadata <- axial_asv %>% 
  left_join(tax) %>%
  separate(Taxon, c("Domain", "Supergroup",
                    "Phylum", "Class", "Order",
                    "Family", "Genus", "Species"), sep = ";", remove = FALSE) %>% 
  left_join(metadata)


##Import as phyloseq object

###Axial data
tax_matrix <- axial_asv_tax_metadata %>% 
  select(Feature.ID, Taxon) %>% 
  distinct() %>% 
  separate(Taxon, c("Domain", "Supergroup", 
                    "Phylum", "Class", "Order",
                    "Family", "Genus", "Species"), sep = ";", remove = FALSE) %>% 
  column_to_rownames(var = "Feature.ID") %>% 
  as.matrix

asv_matrix <- axial_asv_tax_metadata %>% 
  select(Feature.ID, sample, value) %>% 
  pivot_wider(names_from = "sample", values_fill = 0, values_from = value) %>% 
  column_to_rownames(var = "Feature.ID") %>% 
  as.matrix

#Align row names for each matrix
rownames(tax_matrix) <- row.names(asv_matrix)

#Create a sample or control column
axial_asv_tax_metadata <- axial_asv_tax_metadata %>% 
  mutate(sample_or_control = case_when(
    grepl("Control", sample) ~ "control",
    TRUE ~ "sample")) %>% 
  mutate(SampleType = case_when(
    grepl("CTD", sample, ignore.case = TRUE) ~ "Plume",
    grepl("Background", sample, ignore.case = TRUE) ~ "Background",
    grepl("Bsw", sample, ignore.case = TRUE) ~ "Background",
    grepl("Control", sample, ignore.case = TRUE) ~ "Control",
    TRUE ~ "Vent")) 

axial_metadata_cones <- axial_asv_tax_metadata %>% 
  select(sample, sample_or_control, Location_Name, SampleType) %>% 
  distinct() %>% 
  column_to_rownames(var = "sample")

#Import asv and tax matrices
ASV = otu_table(asv_matrix, taxa_are_rows = TRUE)
TAX = tax_table(tax_matrix)
phylo_obj <- phyloseq(ASV, TAX)

# Import metadata as sample data in phyloseq
samplenames <- sample_data(axial_metadata_cones)

# join as phyloseq object
physeq_wnames = merge_phyloseq(phylo_obj, samplenames)

#Identify contaminant ASVs
# When "Control" appears in "Sample_or_Control column, this is a negative control"
sample_data(physeq_wnames)$is.neg <- sample_data(physeq_wnames)$sample_or_control == "control"

#ID contaminants using Prevalence information
contam_prev <- isContaminant(physeq_wnames, 
                             method="prevalence", 
                             neg="is.neg", 
                             threshold = 0.5, normalize = TRUE) 

# Report number of ASVs IDed as contamintants
table(contam_prev$contaminant)
#FALSE 26911 TRUE 648

#0.5 - threshold will ID contaminants in all samples that are more prevalent in negative controls than in positive samples

#Subset contaminant ASVs
contams <- filter(contam_prev, contaminant == TRUE)
list_of_contam_asvs <- as.character(row.names(contams))

taxa_contam <- as.data.frame(tax_matrix) %>% 
  rownames_to_column(var = "Feature.ID") %>% 
  filter(Feature.ID %in% list_of_contam_asvs)

axial_asv_wtax_decon <- axial_asv_tax_metadata %>% 
  filter(!(Feature.ID %in% list_of_contam_asvs)) %>% 
  filter(!(sample_or_control == "control"))

tmp_orig <- (axial_asv_tax_metadata %>% filter(!(sample_or_control == "control")))

# Stats on lost
x <- length(unique(tmp_orig$Feature.ID)); x
#27559

y <- length(unique(axial_asv_wtax_decon$Feature.ID)); y #26911
100*((y-x)/x) #-2.351319
a <- sum(tmp_orig$value);a #8060496
b <- sum(axial_asv_wtax_decon$value);b #6944159
100*((b-a)/a) #-13.84948

## Subsample to clean ASVs
axial_asv_wtax_wstats <- axial_asv_tax_metadata %>% 
  mutate(DECONTAM = case_when(
    Feature.ID %in% list_of_contam_asvs ~ "FAIL",
    TRUE ~ "PASS"
  ))


axial_decontam <- axial_asv_wtax_wstats |> 
  filter(sample_or_control == "sample") |> 
  filter(DECONTAM == "PASS") |> 
  select(Feature.ID, sample, SEQUENCE_COUNT = value, Taxon, Domain,
         Supergroup, Phylum, Class, Order, Family, Genus, Species, Location_Name, 
         SampleType, DEPTH, Prok_min, Prok_mid, Prok_mid2, Prok_max, Prok_avg, POC..um.,
         DOC..um., TDN.um.) |> 
  add_column(dataset = "18S")

axial_decontam_DF <- axial_decontam %>% filter(SEQUENCE_COUNT != 0)

save(axial_decontam_DF, file = "~/r/Axial2023/datainput/Axial_18S_decontam.RData")

##Save as an excel file
load("~/r/Axial2023/datainput/Axial_18S_decontam.RData")
ls()
install.packages("writexl")

library(writexl)
write_xlsx(axial_decontam_DF, "axial_decontam_DF.xlsx")

```
